mmedit/models/backbones/vfi_backbones/tof_vfi_net.py

```diff
@@ -0,0 +1,322 @@
+        """
+        return self.basic_module(tensor_input)
+
+
+class SPyNet(nn.Module):
+    """SPyNet architecture.
+
+    Note that this implementation is specifically for TOFlow. It differs from
+    the common SPyNet in the following aspects:
+        1. The basic modules in paper of TOFlow contain BatchNorm.
+        2. Normalization and denormalization are not done here, as
+            they are done in TOFlow.
+    Paper:
+        Optical Flow Estimation using a Spatial Pyramid Network
+    Code reference:
+        https://github.com/Coldog2333/pytoflow
+
+    Args:
+        norm_cfg (dict | None): Config of normalization.
+        pretrained (str): path for pre-trained SPyNet. Default: None.
+    """
+
+    def __init__(self, norm_cfg, pretrained=None):
+        super().__init__()
+
+        self.basic_module = nn.ModuleList(
+            [BasicModule(norm_cfg) for _ in range(4)])
+
+        if isinstance(pretrained, str):
+            logger = get_root_logger()
+            load_checkpoint(self, pretrained, strict=True, logger=logger)
+        elif pretrained is not None:
+            raise TypeError('[pretrained] should be str or None, '
+                            f'but got {type(pretrained)}.')
+
+    def forward(self, ref, supp):
+        """
+        Args:
+            ref (Tensor): Reference image with shape of (b, 3, h, w).
+            supp: The supporting image to be warped: (b, 3, h, w).
+
+        Returns:
+            Tensor: Estimated optical flow: (b, 2, h, w).
+        """
+
+        num_batches, _, h, w = ref.size()
+        ref = [ref]
+        supp = [supp]
+
+        # generate downsampled frames
+        for _ in range(3):
+            ref.insert(
+                0,
+                F.avg_pool2d(
+                    input=ref[0],
+                    kernel_size=2,
+                    stride=2,
+                    count_include_pad=False))
+            supp.insert(
+                0,
+                F.avg_pool2d(
+                    input=supp[0],
+                    kernel_size=2,
+                    stride=2,
+                    count_include_pad=False))
+
+        # flow computation
+        flow = ref[0].new_zeros(num_batches, 2, h // 16, w // 16)
+        for i in range(4):
+            flow_up = F.interpolate(
+                input=flow,
+                scale_factor=2,
+                mode='bilinear',
+                align_corners=True) * 2.0
+            flow = flow_up + self.basic_module[i](
+                torch.cat([
+                    ref[i],
+                    flow_warp(
+                        supp[i],
+                        flow_up.permute(0, 2, 3, 1),
+                        padding_mode='border'), flow_up
+                ], 1))
+
+        return flow
+
+
+class ResNet(nn.Module):
+    """ResNet architecture.
+
+    Three-layers ResNet/ResBlock
+    reference: https://blog.csdn.net/chenyuping333/article/details/82344334
```

Comment:
As this is not the standard ResNet, the reference link here is actually misleading. 
A better docsting is 
"""A so-called ResNet used in tof-flow

reference: <link to the the original implementation of vfi-tof>
"""